src/controller/proxy/controller.go

```diff
@@ -0,0 +1,230 @@
+	// UseLocalBlob check if the blob should use local
+	UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool
+	// ProxyBlob proxy the blob request to the target server
+	ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error
+	// ProxyManifest proxy the manifest to the target server
+	ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error
+}
+
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	mu          sync.Mutex
+	inflight    map[string]interface{}
+	artifactCtl artifact.Controller
+	local       LocalInterface
+}
+
+func (c *controller) isProxyReady(p *models.Project) bool {
+	if p.RegistryID < 1 {
+		return false
+	}
+	reg, err := c.registryMgr.Get(p.RegistryID)
+	if err != nil {
+		log.Errorf("failed to get registry, error:%v", err)
+		return false
+	}
+	return reg.Status == model.Healthy
+}
+
+func (c *controller) UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	if len(string(art.Digest)) > 0 {
+		exist, err := c.local.BlobExist(ctx, art.Digest)
+		if err == nil && exist {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	exist, err := c.local.BlobExist(ctx, digest)
+	if err != nil {
+		return false
+	}
+	return exist
+}
+
+func (c *controller) ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error {
+	var man distribution.Manifest
+	var err error
+	if len(string(art.Digest)) > 0 {
+		// pull by digest
+		log.Debugf("Getting manifest by digiest %v", art.Digest)
+		man, err = r.ManifestByDigest(repo, string(art.Digest))
+	} else { // pull by tag
+		if len(art.Tag) == 0 {
+			art.Tag = "latest"
+		}
+		log.Debugf("Getting manifest by tag %v", art.Tag)
+		man, err = r.ManifestByTag(repo, string(art.Tag))
+	}
+
+	if err != nil {
+		if errors.IsNotFoundErr(err) && len(art.Tag) > 0 {
+			go func() {
+				c.local.DeleteManifest(ctx, repo, string(art.Tag))
+			}()
+		}
+		serror.SendError(w, err)
+		return err
+	}
+
+	ct, payload, err := man.Payload()
+	setHeaders(w, int64(len(payload)), ct, art.Digest)
+	w.Write(payload)
+
+	// Push manifest in background
+	go func() {
+		c.waitAndPushManifest(ctx, p, repo, string(art.Tag), man, art, ct, r)
+	}()
+
+	return nil
+}
+
+func (c *controller) ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error {
+	log.Debugf("The blob doesn't exist, proxy the request to the target server, url:%v", repo)
+	desc := distribution.Descriptor{}
+	size, bReader, err := r.BlobReader(repo, dig)
+	defer bReader.Close()
```

Comment:
The `defer` should be put after the error checking logic

Reply:
add comment here